### PR TITLE
Flow Definition source added

### PR DESCRIPTION
**Pull request by gautamabhishek46**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -103,6 +103,24 @@
             <version>2.14</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins.workflow</groupId>
+            <artifactId>workflow-cps</artifactId>
+            <version>2.39</version>
+            <scope>compile</scope>
```
Comment (oleg-nenashev) [inline pom.xml]: Why do you use the `compile` scope? You should be using the default scope or `test` for 99% of the use-cases for plugins

File: src/main/java/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 CloudBees, Inc.
```
Comment (oleg-nenashev) [inline src/main/java/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory.java]: If you copy-paste the code, make sure to reference the original source

File: src/main/java/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import java.io.IOException;
+
+/**
+ * Recognizes and builds {@code Jenkinsfile}.
+ */
+public class WorkflowBranchProjectFactory extends org.jenkinsci.plugins.workflow.multibranch.AbstractWorkflowBranchProjectFactory {
```
Comment (oleg-nenashev) [inline src/main/java/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory.java]: Please do not use "Workflow" for any new classes. Should be "Pipeline" if you want to use this term

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        listener.getLogger().format(script);
+        return new CpsFlowExecution(script, false, owner);
+    }
+
+    @Extension
+    public static class DescriptorImpl extends FlowDefinitionDescriptor {
+
+        @Nonnull
+        @Override public String getDisplayName() {
+            return "YAML configuration for multibranch pipeline";
```
Comment (oleg-nenashev) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Better to use localizable strings: https://jenkins.io/doc/developer/internationalization/i18n-source-code/

File: src/main/java/io/jenkins/plugins/sprp/WorkflowMultiBranchProjectFactory.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Extension
+    public static class DescriptorImpl extends MultiBranchProjectFactoryDescriptor {
+
+        @Override public MultiBranchProjectFactory newInstance() {
+            return new org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProjectFactory();
+        }
+
+        @Override public String getDisplayName() {
+            return "Pipeline " + WorkflowBranchProjectFactory.SCRIPT;
```
Comment (oleg-nenashev) [inline src/main/java/io/jenkins/plugins/sprp/WorkflowMultiBranchProjectFactory.java]: Localizable strings: https://jenkins.io/doc/developer/internationalization/i18n-source-code/

File: src/main/java/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory.java (hunk 2 above)
Comment (oleg-nenashev) [reply src/main/java/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory.java]: You also patched the code, so the copyright would need to be updated. E.g. `Copyright 2015-2018 CloudBees, Inc., Abhishek Gautam` or so

File: pom.xml (hunk 1 above)
Comment (gautamabhishek46) [reply pom.xml]: Actually, it was already placed when I created the skeleton.

**Comment by gautamabhishek46** [comment]
Is this good now?

**Comment by oleg-nenashev** [comment]
@gautamabhishek46 Well, we are getting closer. But you have not added your name to license headers of the modified files.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
+                                          List<? extends Action> actions) throws Exception {
+        String script = "node {\n" +
+                        "    stage('Example') {\n" +
+                        "            echo 'Done Cloning'\n" +
+                        "    }\n" +
+                        "}";
```
Comment (grandvizier) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: good use of hardcoding some stuff at the beginning, but are we need already expected this script to be yaml?

Comment (oleg-nenashev) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Yes, there should be some kind of YAML => Pipeline converter finally

File: src/main/java/io/jenkins/plugins/sprp/YAML_BranchProjectFactory.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
+import org.jenkinsci.plugins.workflow.multibranch.AbstractWorkflowBranchProjectFactory;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import java.io.IOException;
+
+/**
+ * Recognizes and builds {@code Jenkins.yaml}.
```
Comment (kwhetstone) [inline src/main/java/io/jenkins/plugins/sprp/YAML_BranchProjectFactory.java]: In keeping with Jenkins naming conventions, I'd suggest calling the file "Jenkinsfile.yaml"

File: src/main/java/io/jenkins/plugins/sprp/YAML_BranchProjectFactory.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+ * Original code: org.jenkinsci.plugins.workflow.multibranch.WorkflowBranchProjectFactory
+ */
+public class YAML_BranchProjectFactory extends AbstractWorkflowBranchProjectFactory {
```
[the remark this answers is not in the document]
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_BranchProjectFactory.java]: I am planing to replace it later.

File: src/main/java/io/jenkins/plugins/sprp/YAML_BranchProjectFactory.java (hunk 7 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_BranchProjectFactory.java]: Ok I will make that change.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 6 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Yes, I will implement this next.

**Comment by kwhetstone** [comment]
Also!  Something else that usually crops up in the Jenkins project: a link back to a Jira ticket if one exists.  You could link this to [JENKINS-51518](https://issues.jenkins-ci.org/browse/JENKINS-51518) in the comments of the PR.  Also there's a reference to the ticket in the title of the Pull Request.  [This](https://github.com/jenkinsci/workflow-multibranch-plugin/pull/63) is a good example.

File: pom.xml (hunk 1 above)
Comment (gautamabhishek46) [reply pom.xml]: should I remove the compile scope ?

File: src/main/resources/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory/config.jelly [hunk 9]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+The MIT License
+
+Copyright 2015 CloudBees, Inc.
```
Comment (oleg-nenashev) [inline src/main/resources/io/jenkins/plugins/sprp/WorkflowBranchProjectFactory/config.jelly]: Not sure it's correct

**Pull request merged**